snuba_sdk/query_validation.py

```diff
@@ -0,0 +1,224 @@
+from __future__ import annotations
+
+from typing import List, Optional, Sequence, Set, Tuple, Union
+
+# Import the modules due to sphinx autodoc problems
+# https://github.com/agronholm/sphinx-autodoc-typehints#dealing-with-circular-imports
+from snuba_sdk import query as main
+from snuba_sdk import query_visitors as qvisitors
+from snuba_sdk.column import Column
+from snuba_sdk.conditions import (
+    BooleanCondition,
+    Condition,
+    Op,
+    get_top_level_conditions,
+)
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import Expression
+from snuba_sdk.function import CurriedFunction
+from snuba_sdk.relationships import Join
+from snuba_sdk.schema import Column as ColumnModel
+
+
+class InvalidMatchError(Exception):
+    pass
+
+
+def validate_match(
+    query: main.Query, column_finder: qvisitors.ExpressionSearcher
```

Comment:
Furthermore, can this just be a method on Query?

Reply:
I pass it in just to avoid initializing it multiple times.

And while it could be a method of query, I am actively trying to keep all this logic separate to avoid making that class more bloated than it already is.

Reply:
> I pass it in just to avoid initializing it multiple times.

ok so why not do this? This way you'll only instantiate it the first time

```python3
def validate_match(
    query: main.Query, column_finder: Optional[qvisitors.ExpressionSearcher] = None
) -> None:
    column_finder = column_finder or ExpressionFinder(Column)
    all_columns = column_finder.visit(query)
```

Reply:
Also I don't think putting this into a separate file makes the codebase less bloated 

this code here really seems like it could just be methods on those specific classes. 

```python3
    if isinstance(query.match, main.Query):
        validate_subquery(query.match, all_columns)
        validate_match(query.match, column_finder)
    elif isinstance(query.match, Join):
        validate_join(query.match, all_columns)
```

Reply:
It doesn't make the code less bloated, it keeps the `Query` class from becoming too large. The pattern of this code base to date has been using the `Visitor` pattern from Snuba, where the code that defines what a Query is is in the class, and the code that traverses a query and validates it is in a separate visitor (in this case in `query_visitors.py`. I originally had all this logic inline in the `visit_match` function of the visitor, but it just became too much code in one function. So I moved it to this separate file. It also made testing a little easier by having distinct functions I could call.

Reply:
> ok so why not do this? This way you'll only instantiate it the first time

I don't see the difference. Both ways instantiate it once and both take the finder as a parameter.

Reply:
> I don't see the difference. Both ways instantiate it once and both take the finder as a parameter.

In the way I suggested, the caller doesn't have to worry about what kind of `ExpressionSearcher` they have to pass in, it's already encapsulated within this module. In the way it is now, there are two things you need to pass in to the `validate_match` function:

1. The query: this makes sense because that's what's being validated.
2. A column finder: which the user of the function now has to figure out what that is. 

if the default argument already gives you what you're supposed to pass in, the user of the function just has to do:

```python3
validate_match(my_query)
```

and you've reduced the mental burden on the user of your API.

You could go a step further and do:

```python3

def _validate_match_impl(query: main.Query, column_finder: qvisitors.ExpressionSearcher):
    """ your function as is"""

def validate_match(query: main.Query):
    column_finder = ExpressionSearcher(Column)
    _validate_match_impl(query, column_finder)
```

But then you cut off the ability for custom column finders (in case you want that).